components/log-ingestor/src/compression/listener.rs

```diff
@@ -0,0 +1,151 @@
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    ///
+    /// * Forwards [`Buffer::add`]'s return values on failure.
+    /// * Forwards [`Buffer::submit`]'s return values on failure.
+    pub async fn run(mut self, cancel_token: CancellationToken) -> Result<()> {
+        let mut timer: Pin<Box<Sleep>> = Box::pin(sleep_until(Instant::now() + self.timeout));
+
+        loop {
+            select! {
+                // Cancellation requested.
+                () = cancel_token.cancelled() => {
+                    // TODO: Log the cancellation event when the logger has been integrated.
+                    self.buffer.submit().await?;
+                    return Ok(());
+                }
+
+                // New object metadata received.
+                optional_object_metadata = self.receiver.recv() => {
+                    match optional_object_metadata {
+                        None => {
+                            self.buffer.submit().await?;
+                            return Err(
+                                anyhow::anyhow!("Listener channel has been closed unexpectedly")
+                            );
+                        }
+                        Some(object_metadata) => {
+                            self.buffer.add(object_metadata).await?;
+                        }
+                    }
+                }
+
+                // Timer fired.
+                () = &mut timer => {
+                    self.buffer.submit().await?;
+                }
+            }
+
+            timer.as_mut().reset(Instant::now() + self.timeout);
+        }
+    }
+}
+
+/// Represents a listener that accepts S3 object metadata from multiple senders and buffers them
+/// for submission.
+pub struct Listener {
+    sender: mpsc::Sender<ObjectMetadata>,
+    cancel_token: CancellationToken,
+    handle: tokio::task::JoinHandle<Result<()>>,
+}
+
+impl Listener {
+    /// Create and spawn a new [`Listener`] backed by a [`ListenerTask`].
+    ///
+    /// This function spawns a [`ListenerTask`]. The spawned task will buffer incoming
+    /// [`ObjectMetadata`] values and call the supplied `Submitter` when either the buffer's
+    /// threshold is reached or the configured `timeout` fires.
+    ///
+    /// # Type parameters
+    ///
+    /// - `Submitter`: A type that implements the [`BufferSubmitter`] trait to submit buffered
+    ///   object metadata.
+    ///
+    /// # Returns
+    ///
+    /// A newly created instance of [`Listener`].
+    pub fn spawn<Submitter: BufferSubmitter + Send + 'static>(
+        buffer: Buffer<Submitter>,
+        timeout: Duration,
+        channel_capacity: usize,
+    ) -> Self {
+        let (sender, receiver) = mpsc::channel(channel_capacity);
+        let task = ListenerTask {
+            buffer,
+            timeout,
+            receiver,
+        };
+        let cancel_token = CancellationToken::new();
+        let child_cancel_token = cancel_token.child_token();
```

Comment:
```suggestion
        let child_cancel_token = cancel_token.clone();
```
I see... Will this work? If child_token() is cancelled, the parent token will not be cancelled. This doesn't sound like a feature we need.

Reply:
thx. Good catch.